Add rule and autofix to sort the contents of `__all__`

AlexWaygood opened the pull request:
## Summary

This implements the rule proposed in #1198 (though it doesn't close the issue, as there are some open questions about configuration that might merit some further discussion). This ended up being a bigger PR than I expected! I've tried to make sure it's well commented, so hopefully it's not too hard to figure out what's going on.

## Test Plan

`cargo test` / `cargo insta review`. I also ran this rule on the CPython codebase, and it produces this diff, which looks pretty good to me: [cpython_diff.txt](https://github.com/astral-sh/ruff/files/13906504/ruff_diff.txt)

AlexWaygood commented:
Some miscellaneous notes:
- We don't try to sort `__all__` if there's any implicitly concatenated strings inside `__all__`.
- We don't try to sort `__all__` if there are any parenthesized items inside `__all__`, e.g.

  ```py
  __all__ = (
      "foo",
      (
          "why_would_you_do_this?",
      ),
      "bar"
  )
  ```

- There's no attempt made to deduplicate elements in `__all__`: elements are first sorted alphabetically, then by the order they originally appeared in.
- Some people on the issue wanted to also be able to sort elements by definition order as well as by alphabetical order. That could possibly be implemented as a followup, but I'm not sure it's worth it. It feels like it would add significant complexity, and I don't know how many people would use it.
- The autofix makes a complete hash of `__all__` definitions like [this](https://github.com/python/cpython/blob/0d8fec79ca30e67870752c6ad4e299f591271e69/Lib/_pydecimal.py#L115-L148) or [this](https://github.com/python/cpython/blob/0d8fec79ca30e67870752c6ad4e299f591271e69/Lib/typing.py#L43-L157), which use comments to create subsections in `__all__`. This is unfortunately reasonably common. Ideally, we'd have an option where users could tell ruff to treat all comments on their own line as section delimiters -- if that option was selected, ruff would sort `__all__` within sections, but wouldn't ever move an element out of the section it's already in. I don't attempt to do that here, however, and there are some open design questions about how we would let users opt into that alternative behaviour (see discussion in https://github.com/astral-sh/ruff/issues/1198).

AlexWaygood commented:
Oh, one other note: I decided to implement this as an RUF rule, since:

1. Although isort has this functionality, it's undocumented, and has always been undocumented
2. It doesn't really feel like it has much to do with imports

Since starting working on the rule, however, I realised that the https://github.com/cpendery/asort tool _does_ also exist, so possibly we could call this rule ASORT001 instead of RUF022. I don't have a strong opinion either way!

zanieb commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,453 @@
+    }
+}
+
+struct SortedDunderAll {
+    was_already_sorted: bool,
+    new_dunder_all: Option<String>,
+}
```
This could be an `enum` e.g. (not necessarily the best names)

```rust
enum SortedDunderAll {
    AlreadySorted,
    Sorted(Option<String>)
}
```

AlexWaygood replied:
Oh nice! Will make that change

charliermarsh replied:
Alternatively, `construct_sorted_all` could return `Option<SortedDunderAll>`, where `None` indicates "already sorted". It's the same idea. `enum SortedDunderAll` is arguably more explicit.

AlexWaygood replied:
There's three states we need to represent here:
- `__all__` was already sorted; don't emit the violation
- `__all__` was not sorted, so we should emit the violation, but we don't know how to safely fix it here for whatever reason; something failed along the way
- `__all__` was not sorted, and here's the fix

For now I've gone with the enum, which I think is definitely better than what I originally had, but the naming still doesn't feel _ideal_. Open to suggestions :)

charliermarsh replied:
Oops, good call!

AlexWaygood replied:
Added some more docs on this in my latest push, so I'll mark this as "resolved" for now, but still open to better naming suggestions!

charliermarsh commented on crates/ruff_linter/src/checkers/ast/analyze/statement.rs:
```diff
@@ -19,6 +19,9 @@ use crate::settings::types::PythonVersion;
 pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
+    if checker.settings.rules.enabled(Rule::UnsortedDunderAll) {
+        ruff::rules::sort_dunder_all(checker, stmt);
+    }
```
I'd recommend adding separate methods in `sort_dunder_all` for `Assign` and `AugAssign` (and probably also `AnnAssign`?), and then have those call into a single method internally that takes the statement and value. That way, we ensure we're only calling this rule on the correct AST nodes.

AlexWaygood replied:
I wondered what you might say about this bit :)

Just to double-check I understand: I should have three separate functions in `sort_dunder_all.rs`, with these signatures?

```rs
pub(crate) fn sort_dunder_all_assign(checker: &mut Checker, stmt: &ast::StmtAssign) {}
pub(crate) fn sort_dunder_all_annassign(checker: &mut Checker, stmt: &ast::StmtAnnAssign) {}
pub(crate) fn sort_dunder_all_assign(checker: &mut Checker, stmt: &ast::StmtAugAssign) {}
```

And then call out to those from different places in `checkers/ast/analyze/statement.rs`?

BurntSushi replied:
@charliermarsh Is this how the existing lints work? It is interesting to me because it seems nice for the lint implementation itself to own when it is applicable. Otherwise, you wind up redoing the case analysis. Although I suppose if _all_ lints did this, then every lint would do its own case analysis and maybe that would be bad for perf reasons. I think with your suggested approach, maybe most lints don't need to do any case analysis.

@AlexWaygood Your code snippet looks right to me. Although I do wonder if you could avoid re-doing case analysis on `ast::Stmt`. Basically, have your internal function accept a `target` and `original_value`. Not 100% sure on that though.

AlexWaygood replied:
> Although I do wonder if you could avoid re-doing case analysis on `ast::Stmt`. Basically, have your internal function accept a `target` and `original_value`. Not 100% sure on that though.

Charlie [called me out on that](https://github.com/astral-sh/ruff/pull/9313#discussion_r1438924016) last time I tried to do something like that 😅

BurntSushi replied:
Right. I don't mean that. I'm still proposing that you have the three functions @charliermarsh suggested above as the _public_ API for your new lint module. But internally, instead of re-materializing an `ast::Stmt` and just pattern matching on it again, you extract the state you need and pass it to your main internal routine that implements the lint.

AlexWaygood replied:
Ah, I think we're on the same page! I just pushed cc42bf9 -- is that what you had in mind?

BurntSushi replied:
Precisely. :-)

BurntSushi commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,455 @@
+    Sorted(Option<String>),
+}
+
+fn collect_dunder_all_lines(range: TextRange, locator: &Locator) -> Option<Vec<DunderAllLine>> {
```
I also think some comments describing what this function does would be helpful. It seems like a very crucial detail of how  this lint works. In particular, it is the part that tokenizes a region of the code that we already have an AST for. I think I can make some educated guesses as to why you're doing that here (I'm still somewhat new to this code too), but a comment explaining/motivating the technique would be quite helpful I think!

BurntSushi commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,455 @@
+    checker.diagnostics.push(diagnostic);
+}
+
+struct DunderAllValue<'a> {
```
And I'll say that it isn't obvious to me that this invariant holds, but I think, for example, `construct_sorted_all` depends on it (it will panic, I believe, if `self.items` is empty). I was thinking this invariant was upheld by `if elts.len() < 2 { return None }` below. But `elts` gets passed through `collect_dunder_all_lines` and then `collect_dunder_all_items`. And it's the result of that that is used for `DunderAllValue::items`. But is the length preserved through those operations? I'm not sure. (I mused about this below too.)

BurntSushi reviewed (commented):
Nice work! I think some extra docs on some of the more complicated helper functions/types would be beneficial here. There are some interesting invariants at play here. :-)

AlexWaygood commented:
Thanks for the extremely helpful review @BurntSushi!

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,455 @@
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
+            // Should be safe: we should already have returned by now if there are 0 items
```
I had a go at revamping the comments in this function in https://github.com/astral-sh/ruff/pull/9474/commits/454e7301a15e4b3c05bb584338e711ec6e8d05dc. The most important thing is that, while I think you're right that we probably don't have this function called at all unless `__all__` has at least two items, there's an even stronger reason to know that this invariant holds true: if `__all__` has one item or less, then `self.items` and `sorted_items` _have_ to compare equal, so we short-circuit here, and can safely rely on both of them having length >=2 for the rest of the function: https://github.com/astral-sh/ruff/pull/9474/files#diff-09e6939507231a0c2052ea73730cfbc0045bc773134683b2176724a3e9acee75R232-R235

BurntSushi replied:
I think if some part of this comment made its way into the source code as a comment, that would be great. :-) Thank you for digging in!

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+
+struct DunderAllValue<'a> {
+    items: Vec<DunderAllItem>,
+    range: &'a TextRange,
```
This is a good instinct, but `TextRange` is small and implements `Copy`, so it's preferable to just use `TextRange` here. (You won't even need to `.clone()` it.)

AlexWaygood replied:
I keep on thinking today will be the day I get to use lifetimes properly for the first time 🙃

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
```
Can you say more about the bracket handling here? I'm trying to follow. Typically, we do something more like this, where we track all open and closing brackets via a counter, to know if we're at the top level: https://github.com/astral-sh/ruff/blob/eb4ed2471b48d6ccba74064c65e37e32cba41a6c/crates/ruff_python_parser/src/soft_keywords.rs#L67. In this case, we're aborting if we see nested brackets, is that right?

AlexWaygood replied:
> In this case, we're aborting if we see nested brackets, is that right?

Yup, that's correct. If there are any parenthesized items in `__all__`, that would required a lot more complexity to handle, and it seems like a massive edge case, so not worth it.

> should `Lbrace` also be included?

It's currently handled by the catch-all `_` condition at the end of the `match` (which just returns `None` if we encounter any token not explicitly enumerated already), but I can move it here if you like?

charliermarsh replied:
In that case, though, why  is `Lsqb` handled here?

AlexWaygood replied:
Because `__all__` could be a list, and we don't want to reject the opening square bracket of `__all__` if it's a list

AlexWaygood replied:
Added a big comment clarifying this in 572bb66546b9d0307b0de8e1e576041890410396 :)

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+    let mut this_range = None;
+    let mut idx = 0;
+    for line in lines {
+        let DunderAllLine { items, comment } = line;
```
Nit: consider destructuring in the iterator? Like `for DunderAllLine { ... } in lines { ... }`? Personal preference though, not blocking.

AlexWaygood replied:
ooh you can do that? did not know

BurntSushi replied:
The terminology here is that irrefutable patterns are accepted wherever names are assigned. This even includes function parameters!

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+                idx += 1;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
```
I feel like it should be possible to avoid this clone, since we don't need `lines` beyond this method. We could consider either making `lines` an owned `Vec` to this method, or using lifetimes on `DunderAllItem`, and having it take `&str` that's tied to the lifetime of `lines: &[DunderAllLine]`. (I'd probably do the latter.)

AlexWaygood replied:
I think I initially tried doing the latter, but the borrow checker was very angry with me, so I ran away. I'll have another go tomorrow.

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+) -> Option<String> {
+    let Some(indent) = indentation(locator, parent) else {
+        return None;
+    };
```
Nit: this can be `let indent = indentation(locator, parent)?;`

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            }
```
I tend to think that using unsafe to protect comments is a bit of a misuse of the concept, which is meant to protect against possible breakages via changes in semantics. But, this question has come up before, and I don't know that we have a unified answer.

\cc @zanieb

AlexWaygood replied:
I feel like as a _user_ of ruff I'd probably be a little annoyed if ruff did something like https://github.com/astral-sh/ruff/pull/9474#issuecomment-1888091353 to my carefully designed `__all__` definition, and then claimed it was a "safe" fix 😄 but I'll defer to you and Zanie :)

MichaReiser replied:
That's a fair point. We would likely need to mark all fixes as unsafe following that reasoning because:
* some fixes don't handle comments at all
* Many fixes don't produce nicely formatted code. You could use the same argument that we should use unsafe because ruff could mess up my carefully hand-formatted code. 

Let's see what @zanieb thinks

AlexWaygood replied:
I've pushed 24616f3a55e18640c92b548177842ba65021bd9c to mark the fix as always being safe, but I've kept a note in the docs stating that it might sometimes move comments to unexpected locations

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,499 @@
+            SortedDunderAll::Sorted(value) => value,
+        };
+
+    let dunder_all_range = dunder_all_val.range();
```
Personal style, I'd probably just inline this wherever it's used, I don't know that the binding adds much but it does add more things named `dunder_all_*` to worry about.

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+        })
+    }
+
+    fn construct_sorted_all(
```
Nit: I think idiomatically in Rust you might call this `as_sorted()` (if it takes `&self`) or `into_sorted()`. In this case, I think `into_sorted` would make sense, then it could take `self` instead of `&self` and consume `self.items` (so no need to `.clone()`).

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+                    comment_in_line = None;
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
```
Can you walk me through how the comment-handling works here? How are own-line comments handled? (It looks like `comment_in_line` is for end-of-line comments, like `"foo" # bar`.) What happens if there are multiple own-lines comments between items?

charliermarsh replied:
Oh, a `DunderAllLine` can be _just_ a comment, with no item.

AlexWaygood replied:
Yeah, at this stage we're not trying to decide which elements "own" which comments, we're just trying to determine "on line 0 there's a comment"; "on line 1 there's two items and a comment"; "one line 2 there's an item and no comment", etc.

Later, when we pass the list of `DunderAllLine`s to `collect_dunder_all_items()`, we try to determine which elements in `__all__` "own" which comments (meaning the comments should move with the element when the element moves as part of being sorted).

charliermarsh replied:
For clarity, I'm wondering if we should use an enum to represent "line with items, and optionally an end-of-line comment" vs. "own-line comment"? It took me a bit of time to get there in understanding that the struct was being repurposed in this way.

AlexWaygood replied:
> For clarity, I'm wondering if we should use an enum to represent "line with items, and optionally an end-of-line comment" vs. "own-line comment"? It took me a bit of time to get there in understanding that the struct was being repurposed in this way.

That's a really great suggestion -- tried it out, and it makes the code much clearer. Will push shortly.

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+                        range.start(),
+                    ));
+                }
+                break;
```
Even though we're breaking, it might be worth adding:

```rust
                    items_in_line.clear();
                    comment_in_line = None;
```

...to the end of the `if` block above, as a consistent "reset condition".

AlexWaygood replied:
I get this warning if I try to do that:

```
value assigned to `comment_in_line` is never read
maybe it is overwritten before being read?
`#[warn(unused_assignments)]` on by default
```

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
```
Nit: I have a preference towards avoiding single-letter variables in favor of more explicit names. E.g., `item` and `range` here, maybe?

charliermarsh replied:
I think we should be able to avoid this `.to_owned()`... E.g., could we pass in `&mut Vec`? Or maybe we pass in `items.drain(..)` as the argument?

AlexWaygood replied:
@MichaReiser and I vanquished all uses of `.clone()` in a pair-programming session; this region of code no longer exists :)

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
+                        original_index: idx,
```
Could `idx` here just be `all_items.len()`, instead of tracking separately?

AlexWaygood replied:
done!

charliermarsh commented on crates/ruff_linter/src/checkers/ast/analyze/statement.rs:
```diff
@@ -1065,12 +1065,15 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
             }
         }
-        Stmt::AugAssign(ast::StmtAugAssign { target, .. }) => {
+        Stmt::AugAssign(augassign @ ast::StmtAugAssign { target, .. }) => {
```
Nit: `aug_assign` to match the `AugAssign` casing? Same with the method name.

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,455 @@
+    // We're only interested in `__all__ = ...` and `__all__ += ...`
+    let (target, original_value) = match stmt {
+        ast::Stmt::Assign(ast::StmtAssign { value, targets, .. }) => match targets.as_slice() {
+            [ast::Expr::Name(ast::ExprName { id, .. })] => (id, value.as_ref()),
```
I always use `.as_ref()` for this, FWIW. I might be doing it wrong then :)

BurntSushi replied:
Yeah that's the evil thing about `AsRef::as_ref()`. It works, and then you upgrade Rust, and maybe it breaks. We haven't gotten one in a while, but there have been lots of reports of breakage in the past because std added a trait impl somewhere.

(This is an instance of a more general problem where adding trait impls can cause inference to fail where it previously succeeded. `AsRef` just happens to be one particular instance of it.)

With all that said, the worst thing that happens is our code will stop compiling when a new version of Rust is released. And a fix is (guaranteed actually) to be simple, so long as you diagnose it correctly. You can never get something wrong. But... these things stand out to me because I've been bitten by it myself and users reporting breakage to me have been bitten by it.

AlexWaygood replied:
I had a bit of trouble locally getting it to work without `.as_ref()` :(

@burntsushi, would you mind pasting a diff of what you mean here? Or feel free to push straight to my branch if you prefer, whichever's easiest :)

BurntSushi replied:
Yeah! So it looks like there are only two `.as_ref()` calls left here. This diff is the simplest one that removes them:

```
[andrew@duff ruff]$ g diff
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
index d42249f38..227454cea 100644
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
@@ -101,7 +101,7 @@ pub(crate) fn sort_dunder_all_aug_assign(
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = &**target else {
         return;
     };
     sort_dunder_all(checker, id, value, parent);
@@ -122,7 +122,7 @@ pub(crate) fn sort_dunder_all_ann_assign(
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = &**target else {
         return;
     };
     sort_dunder_all(checker, id, val, parent);
```

The key here is noticing that `target` has type `&Box<Expr>`. Because you can't pattern match through a `Box` (a limitation that everyone would like to lift, but is I guess difficult to do so), and since you want to borrow from the AST, you want a `&Expr` here. To go from a `&Box<Expr>`, you can dereference twice to get an `Expr` (temporarily), and then add a `&` to get a `&Expr`.

Now for me personally, this is not how I would write this. The only reason why it works is something called [match ergonomics](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html) that essentially tries to remove the need to explicitly mark whether a pattern match is borrowed or not. I think match ergonomics makes code like this less clear because it's harder to tell at a glance what `target` is just from reading the source code.

This is how I would write it:

```
[andrew@duff ruff]$ g diff
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
index d42249f38..ed3ee0340 100644
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
@@ -93,15 +93,15 @@ pub(crate) fn sort_dunder_all_aug_assign(
     parent: &ast::Stmt,
 ) {
     let ast::StmtAugAssign {
-        value,
-        target,
+        ref value,
+        ref target,
         op: ast::Operator::Add,
         ..
-    } = node
+    } = *node
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { ref id, .. }) = **target else {
         return;
     };
     sort_dunder_all(checker, id, value, parent);
@@ -115,14 +115,14 @@ pub(crate) fn sort_dunder_all_ann_assign(
     parent: &ast::Stmt,
 ) {
     let ast::StmtAnnAssign {
-        target,
-        value: Some(val),
+        ref target,
+        value: Some(ref val),
         ..
     } = node
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { ref id, .. }) = **target else {
         return;
     };
     sort_dunder_all(checker, id, val, parent);
```

Now you still need the `**target`, but the fact that you need it is now (maybe) a little clearer. That match on `ref target` tells you that `target` _has_ to be a `&SOMETHING`.

AlexWaygood replied:
Thanks so much <3. I applied your suggestion in https://github.com/astral-sh/ruff/pull/9474/commits/229fec17edf6aa1b1ed2c2a365b9e81e740908db.

I _think_ I understand the solution, but will try to read up a bit more this weekend to check I fully understand what's going on here :)

BurntSushi replied:
Yeah it will take some getting used to. There's a reason why match ergonomics exists even if I'm no fan. :-)

charliermarsh commented:
@AlexWaygood -- Given:

```python
__all__ = [
    # Two major classes
    'Decimal', 'Context',

    # Named tuple representation
    'DecimalTuple',

    # Contexts
    'DefaultContext', 'BasicContext', 'ExtendedContext',

    # Exceptions
    'DecimalException', 'Clamped', 'InvalidOperation', 'DivisionByZero',
    'Inexact', 'Rounded', 'Subnormal', 'Overflow', 'Underflow',
    'FloatOperation',

    # Exceptional conditions that trigger InvalidOperation
    'DivisionImpossible', 'InvalidContext', 'ConversionSyntax', 'DivisionUndefined',

    # Constants for use in setting up contexts
    'ROUND_DOWN', 'ROUND_HALF_UP', 'ROUND_HALF_EVEN', 'ROUND_CEILING',
    'ROUND_FLOOR', 'ROUND_UP', 'ROUND_HALF_DOWN', 'ROUND_05UP',

    # Functions for manipulating contexts
    'setcontext', 'getcontext', 'localcontext',

    # Limits for the C version for compatibility
    'MAX_PREC',  'MAX_EMAX', 'MIN_EMIN', 'MIN_ETINY',

    # C version: compile time choice that enables the thread local context (deprecated, now always true)
    'HAVE_THREADS',

    # C version: compile time choice that enables the coroutine local context
    'HAVE_CONTEXTVAR'
]
```

How exactly does that get formatted after this fix? (Just following up per one of your comments above.)

AlexWaygood commented:
> @AlexWaygood -- Given:
> 
> ```python
> __all__ = [
>     # Two major classes
>     'Decimal', 'Context',
> 
>     # Named tuple representation
>     'DecimalTuple',
> 
>     # Contexts
>     'DefaultContext', 'BasicContext', 'ExtendedContext',
> 
>     # Exceptions
>     'DecimalException', 'Clamped', 'InvalidOperation', 'DivisionByZero',
>     'Inexact', 'Rounded', 'Subnormal', 'Overflow', 'Underflow',
>     'FloatOperation',
> 
>     # Exceptional conditions that trigger InvalidOperation
>     'DivisionImpossible', 'InvalidContext', 'ConversionSyntax', 'DivisionUndefined',
> 
>     # Constants for use in setting up contexts
>     'ROUND_DOWN', 'ROUND_HALF_UP', 'ROUND_HALF_EVEN', 'ROUND_CEILING',
>     'ROUND_FLOOR', 'ROUND_UP', 'ROUND_HALF_DOWN', 'ROUND_05UP',
> 
>     # Functions for manipulating contexts
>     'setcontext', 'getcontext', 'localcontext',
> 
>     # Limits for the C version for compatibility
>     'MAX_PREC',  'MAX_EMAX', 'MIN_EMIN', 'MIN_ETINY',
> 
>     # C version: compile time choice that enables the thread local context (deprecated, now always true)
>     'HAVE_THREADS',
> 
>     # C version: compile time choice that enables the coroutine local context
>     'HAVE_CONTEXTVAR'
> ]
> ```
> 
> How exactly does that get formatted after this fix? (Just following up per one of your comments above.)

Here's the diff:

<details>

```diff
diff --git a/Lib/_pydecimal.py b/Lib/_pydecimal.py
index 2692f2fcba..382f9c4807 100644
--- a/Lib/_pydecimal.py
+++ b/Lib/_pydecimal.py
@@ -113,38 +113,53 @@
 """
 
 __all__ = [
+    'BasicContext',
+    'Clamped',
+    'Context',
+    'ConversionSyntax',
     # Two major classes
-    'Decimal', 'Context',
-
+    'Decimal',
+    # Exceptions
+    'DecimalException',
     # Named tuple representation
     'DecimalTuple',
-
     # Contexts
-    'DefaultContext', 'BasicContext', 'ExtendedContext',
-
-    # Exceptions
-    'DecimalException', 'Clamped', 'InvalidOperation', 'DivisionByZero',
-    'Inexact', 'Rounded', 'Subnormal', 'Overflow', 'Underflow',
-    'FloatOperation',
-
+    'DefaultContext',
+    'DivisionByZero',
     # Exceptional conditions that trigger InvalidOperation
-    'DivisionImpossible', 'InvalidContext', 'ConversionSyntax', 'DivisionUndefined',
-
-    # Constants for use in setting up contexts
-    'ROUND_DOWN', 'ROUND_HALF_UP', 'ROUND_HALF_EVEN', 'ROUND_CEILING',
-    'ROUND_FLOOR', 'ROUND_UP', 'ROUND_HALF_DOWN', 'ROUND_05UP',
-
-    # Functions for manipulating contexts
-    'setcontext', 'getcontext', 'localcontext',
-
-    # Limits for the C version for compatibility
-    'MAX_PREC',  'MAX_EMAX', 'MIN_EMIN', 'MIN_ETINY',
-
+    'DivisionImpossible',
+    'DivisionUndefined',
+    'ExtendedContext',
+    'FloatOperation',
+    # C version: compile time choice that enables the coroutine local context
+    'HAVE_CONTEXTVAR',
     # C version: compile time choice that enables the thread local context (deprecated, now always true)
     'HAVE_THREADS',
-
-    # C version: compile time choice that enables the coroutine local context
-    'HAVE_CONTEXTVAR'
+    'Inexact',
+    'InvalidContext',
+    'InvalidOperation',
+    'MAX_EMAX',
+    # Limits for the C version for compatibility
+    'MAX_PREC',
+    'MIN_EMIN',
+    'MIN_ETINY',
+    'Overflow',
+    'ROUND_05UP',
+    'ROUND_CEILING',
+    # Constants for use in setting up contexts
+    'ROUND_DOWN',
+    'ROUND_FLOOR',
+    'ROUND_HALF_DOWN',
+    'ROUND_HALF_EVEN',
+    'ROUND_HALF_UP',
+    'ROUND_UP',
+    'Rounded',
+    'Subnormal',
+    'Underflow',
+    'getcontext',
+    'localcontext',
+    # Functions for manipulating contexts
+    'setcontext',
 ]
```

</details>

It's all beautifully alphabetical, and the comments are all preserved, but it's probably not what the original author of that code wanted to happen!

charliermarsh commented:
Oh I see, so we're not preserving the "sections" that they've created in their comments -- got it. Agree we could figure out some better solution there (but doesn't need to be in this PR).

MichaReiser commented:
> Oh I see, so we're not preserving the "sections" that they've created in their comments -- got it. Agree we could figure out some better solution there (but doesn't need to be in this PR).

Your example makes me wonder if we could use empty lines to recognize sections (regardless if there's a comment or not). 

Doing so has the added benefit that we preserve empty lines between imports (which is probably something that was intended).

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+/// ## Fix safety
+/// This rule's fix should be safe for single-line `__all__` definitions
+/// and for multiline `__all__` definitions without comments.
+/// For multiline `__all__` definitions that include comments,
+/// the fix is marked as unsafe, as it can be hard to tell where the comments
+/// should be moved to when sorting the contents of `__all__`.
```
This is an intereseting use of unsafe, but it fits `unsafe`'s definition.

> The fix may be what the user intended, but it is uncertain; the resulting code will have valid syntax.

I would probably keep the fix as safe if we feel like it does the right thing in the majority of cases even if there are comments. But I'm also okay going wiht unsafe (can also be easily changed later). Interested to hear what other people think.

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+                        range.start(),
+                    ));
+                    items_in_line.clear();
+                    comment_in_line = None;
```
You might be able to avoid cloning the value inside of `DunderAllLine` by using `std::mem::take`. `std::mem::take` *takes* the value from `items_in_line` and replaces it with the `Default` (an empty list in that case). 
```suggestion
                    lines.push(DunderAllLine::new(
                        std::mem::take(&mut items_in_line),
                        comment_in_line,
                        range.start(),
                    ));
                    comment_in_line = None;
```

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+        parent: &ast::Stmt,
+        stylist: &Stylist,
+    ) -> SortedDunderAll {
+        let mut sorted_items = self.items.clone();
+        sorted_items.sort();
+
+        // As well as saving us unnecessary work,
+        // returning early here also means that we can rely on the invariant
+        // throughout the rest of this function that both `items` and `sorted_items`
+        // have length of at least two. If there are fewer than two items in `__all__`,
+        // it is impossible for them *not* to compare equal here:
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
```
You can use `self.items.is_sorted()` to test if the items are sorted (and avoids cloning all elements). This has the advantage that checking if the items are sorted is `O(n)` instead of `O(n log(n))` (sorting) + `O(n)` (comparing). The complexity remains the same if the items aren't sorted because you have to call `self.items.sort()` to sort the items but that should be less common for projects actively using the lint rule.

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+        }
+        assert!(self.items.len() >= 2);
+
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis
+        //    (if there was one), up to the start of the first element in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
```
Nit: One thing I like to do is to add a python snippet to give readers a concrete example:

```suggestion
        // As well as the "items" in the `__all__` definition,
        // there is also a "prelude" and a "postlude":
        //  - Prelude == the region of source code from the opening parenthesis
        //    (if there was one), up to the start of the first element in `__all__`.
        //  - Postlude == the region of source code from the end of the last
        //    element in `__all__` up to and including the closing parenthesis
        //    (if there was one).
        // 
        // ```python
        // __all__ = [
        //   "first item",
        //   "last item"   
        // ]
        // ```
        // The prelude in the above example is the text from right after the `[` up to `"first item"`. The postlude is the text between `"last item"` and `]`.
        let prelude_end = {
```

AlexWaygood replied:
I added a variant of this in 5e129ade8e77090a2aa1017611258b2c2f65d942. (By the way: writing that up was what made me spot the bug regarding multiline comments that I fixed in fd1dc54c80004d1b2377ca93ad30718d83e77587!)

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
+        assert!(self.items.len() >= 2);
```
Nit: You can avoid some of the `expect` messages below by doing

```rust
let [first, .., last] = self.items.as_slice() else {
    panic!("Expected at least two items")
};
```

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+                first_item_line_offset
+            }
+        };
+        let (needs_trailing_comma, postlude_start) = {
```
Do I understand this correctly (without having read through all the code) that the fix inserts a trailing comma if the last item is on its own line? 

What about users that prefer not to have trailing commas? Should we instead detect if the source contains a trailing comma and preserve it?

AlexWaygood replied:
> Do I understand this correctly (without having read through all the code) that the fix inserts a trailing comma if the last item is on its own line?

yes, that's correct

> What about users that prefer not to have trailing commas? Should we instead detect if the source contains a trailing comma and preserve it?

good idea; I'll try to implement that

AlexWaygood replied:
Tweaked it in 6f5f55843c7bfc3ea3f7a336cacfceb33eeeb9c3: we now do not add a trailing comma if there wasn't one in the original source!

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+                (true, last_item_line_offset)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
```
We can avoid allocating here because you're constructing a new string below anyway. Sorry, I'm very picky when it comes to allocations because they tend to be expensive (see https://vimeo.com/649009599)


```suggestion
        let mut prelude = Cow::Borrowed(locator.slice(TextRange::new(self.start(), prelude_end)));

        let joined_items = if self.multiline {
            let indentation = stylist.indentation();
            let newline = stylist.line_ending().as_str();
            prelude = Cow::Owned(format!("{}{}", prelude.trim_end(), newline));
            join_multiline_dunder_all_items(
                &sorted_items,
                locator,
                parent,
                indentation,
                newline,
                needs_trailing_comma,
            )
        } else {
            Some(join_singleline_dunder_all_items(&sorted_items, locator))
        };

        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));

        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
        SortedDunderAll::Sorted(new_dunder_all)
    }
```

AlexWaygood replied:
Should `postlude` also be a `Cow`, or just `prelude`?

AlexWaygood replied:
Having done a bit more reading, I can answer my own question: no, `postlude` doesn't need to be a `Cow`, since the type of `postlude` currently is already a borrowed type (`&str`), which is what we want, but the type of `prelude` currently is an owned type (`String`), which is what we don't want, and what using `Cow` fixes for us.

Thanks!

AlexWaygood replied:
bec4a1b714f53fce67b98acaae7b922934537cc2

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+                items,
+                comment_range,
+            }) => {
+                let [(first_val, first_range), rest @ ..] = items.as_slice() else {
+                    unreachable!(
+                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
+                    )
+                };
```
We can avoid cloning `first_val` and `value` below by passing the owned `Vec<DunderAllLine>` to `collect_dunder_all_items` (instead of a slice) and use `into_iter` here that gives you owned values:


```suggestion
                let mut iter = items.into_iter();
                let Some((first_val, first_range)) = iter.next() else {
                    unreachable!(
                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
                    )
                };

                let rest = iter;
```

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+    for line in lines {
+        match line {
+            DunderAllLine::JustAComment(LineWithJustAComment(comment_range)) => {
+                this_range = Some(*comment_range);
```
`this_range` seems to be `None` except if it is a comment. How about renaming it `preceding_comment_range`?


What happens if an item has multiple preceding lines? Should this be changed to `this_range.get_or_insert(comment_range)` to only set the range when `this_range` is `None`?

```python
__all__ = [
    "test"
    # comment
    # another comment only line
    "item"
]
```

AlexWaygood replied:
> How about renaming it `preceding_comment_range`?

Sounds good



> What happens if an item has multiple preceding lines? Should this be changed to `this_range.get_or_insert(comment_range)` to only set the range when `this_range` is `None`?

Ah, I noticed this and fixed it before seeing this review comment 😄 see fd1dc54c80004d1b2377ca93ad30718d83e77587

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
```
I think a good approximation of the number of `items` is to use the number of lines if it's a multiline value and the number of items in the first line otherwise.. Initializing items with a specific capacity can help to reduce the number of times the vector needs to be resized.
```suggestion
    let mut all_items = Vec::with_capacity(match lines.as_slice() {
        [DunderAllLine::OneOrMoreItems(single)] => single.items.len(),
        _ => lines.len(),
    });
```

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+    let mut this_range = None;
+    for line in lines {
+        match line {
+            DunderAllLine::JustAComment(LineWithJustAComment(comment_range)) => {
```
What happens with `JustAComment` if it is the last line? 

```python
__all__ = [
    'test',
    # comment
]
```

I'm asking because `this_range` then never gets added to all items. Will it be part of the prelude or is there an invariant that prevents that `LineWithJustAComment` can be the last line?

AlexWaygood replied:
Any comments after the last item are discarded here, but are added back as part of the postlude in `into_sorted_source_code()` (which used to be called `construct_sorted_all()`; I just renamed it)

AlexWaygood replied:
Added some more docs about this; is this clearer now?

MichaReiser replied:
Thanks. That helps

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
+                    )
+                };
+                let range = this_range.map_or(*first_range, |r| {
+                    TextRange::new(r.start(), first_range.end())
+                });
```
A comment here might be helpful explaining why we're extending the range.

*Include the range of any preceding comment in the range of the item to ensure the comment gets moved with the item*.

AlexWaygood replied:
af6ca0556fe8c3906bd5e52b85f0129a3e41fef0

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+                    additional_comments: *comment_range,
+                });
+                this_range = None;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
+                        original_index: all_items.len(),
+                        range: *range,
+                        additional_comments: None,
+                    });
+                }
```
We can use `Vec::extend` here to let `Vec` know that we intend to push multiple items. It can help `Vec` to regrow the vector only once


```suggestion
                all_items.extend(rest.map(|(value, range)| DunderAllItem {
                    value,
                    original_index: all_items.len(),
                    range,
                    additional_comments: None,
                }));
```

AlexWaygood replied:
(For posterity: we tried this out in a pair-programming session, but the borrow checker was very angry, so we reverted it)

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+impl Eq for DunderAllItem {}
+
+impl DunderAllItem {
+    fn sort_index(&self) -> (&str, usize) {
+        (&self.value, self.original_index)
+    }
+}
```
Can you explain (or add a comment) why `sort_index` is needed. My guess is that it is to preserve the original order, but that's something that Rust's `sort` method should already give us for free.

AlexWaygood replied:
yup, my worry here was about the possibility of an `__all__` that had duplicate items in it. I wanted to ensure that in the eventuality where two items in the list compared equal, the item that was originally first in the list remained first in the list. That invariant will hold even if I just use `self.value` as the key for `.sort()`?

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+        if !is_final_item || needs_trailing_comma {
+            new_dunder_all.push(',');
+        }
+        if let Some(comment) = item.additional_comments {
```
Are `additional_comments` always comments on the same line as the item? If so, can we rename it to `end_of_line_comments` or `trailing_comments` to match the naming convention we use in the formatter.

AlexWaygood replied:
9e2bae25679168aa15178f622366dee2481e960c

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,532 @@
+            new_dunder_all.push(',');
+        }
+        if let Some(comment) = item.additional_comments {
+            new_dunder_all.push_str("  ");
```
This might be a bit opinionated for some :laughing:. Any chance we can preserve the original spacing?

AlexWaygood replied:
Ew @ anybody who doesn't put two spaces before an inline comment, but: see 655bf67ae3459da6b0bec34ab959d56cd4d26bfe :)

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+}
+
+fn sort_dunder_all(checker: &mut Checker, target: &str, node: &ast::Expr, parent: &ast::Stmt) {
+    if target != "__all__" {
+        return;
+    }
```
I think the redefinition here overrides the import at the top (or, if it's an import below the assignment, the import would override the assignment). Python just uses the runtime value of `__all__`; it doesn't attempt to read the value statically in any way (though some type checkers _do_ do that). So I don't _think_ this edge case should have any impact on the way this rule works

konstin replied:
Looks like cpython does indeed not special case `__all__`: https://github.com/search?q=repo%3Apython%2Fcpython+__all__+language%3AC+&type=code

AlexWaygood commented:
I discovered a small bug relating to multiline comments -- working on a fix...

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Sort `__all__` alphabetically".to_string())
```
I like that idea!

AlexWaygood replied:
Implemented in eb96b894f5697f6ac3908e8879e3935f7a317a05!

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,495 @@
+    fn from_expr(value: &ast::Expr, locator: &Locator) -> Option<DunderAllValue> {
+        // Step (1): inspect the AST to check that we're looking at something vaguely sane:
+        let is_multiline = locator.contains_line_break(value.range());
+        let (elts, range) = match value {
```
Thanks, addressed in 768a3648636e6abbff2ed3373e686c2678b38dc1!

AlexWaygood commented:
Thanks all for the detailed feedback, really appreciate it! I _think_ I've now either addressed or asked clarification on all comments, but sorry if I missed something :)

AlexWaygood commented:
Recent changes:
- Fixed a bunch of formatting edge cases I spotted while leafing through the diff this PR branch produced when running ruff with `--fix --unsafe-fixes --preview --select=RUF022` on CPython. Here's the new diff on CPython: [cpython_diff.txt](https://github.com/astral-sh/ruff/files/13929051/cpython_diff.txt)
- Switched to an "isort-style" sort instead of a natural sort: SCREAMING_SNAKE_CASE members come first, then CamelCase members, and then anything else after that. Within each category, items are ordered according to a natural sort.

MichaReiser reviewed (changes requested):
Overall this looks good to me, but there are a few clarification that are needed for me to understand the code and:

* We should align the `PartialEq` and `Ord` implementation of `AllDunderItems`
* Let's add some tests around unparenthesized tuples that start with a parenthesized first item (see inline comment). I suspect that things might go wrong in that case


It would be interesting to explore if we can implement the check as an AST rule and only use lexing for generating the fix. I'll leave that up to you if you want to tackle this and if so, if you want to do it as part of this PR.

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,823 @@
+    for pair in lexer::lex_starts_at(locator.slice(range), Mode::Expression, range.start()) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            // If exactly one `Lpar` or `Lsqb` is encountered, that's fine
+            // -- a valid __all__ definition has to be a list or tuple,
+            // and most (though not all) lists/tuples start with either a `(` or a `[`.
+            //
+            // Any more than one `(` or `[` in an `__all__` definition, however,
+            // indicates that we've got something here that's just too complex
+            // for us to handle. Maybe a string element in `__all__` is parenthesized;
+            // maybe the `__all__` definition is in fact invalid syntax;
+            // maybe there's some other thing going on that we haven't anticipated.
+            //
+            // Whatever the case -- if we encounter more than one `(` or `[`,
+            // we evidently don't know what to do here. So just return `None` to
+            // signal failure.
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
+            }
```
> How common are unparenthesized tuples for `__all__`? I thought `__all__` is recommended to be a list. Maybe it's not worth supporting fixing tuples?

Tuple `__all__`s are very common; I'd definitely rather support them here. I don't think there's a recommendation w.r.t. whether they should be tuples or lists -- I'm not actually sure which is more common. Conceptually, I sort-of prefer using a tuple for `__all__`, personally. But I think that's just a vague instinct towards preferring immutable types for global constants, rather than something that's founded on any strong reasoning :)

Using an _unparenthesized_ tuple for `__all__` is less common, but not unheard of:

- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/base_events.py#L52
- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/coroutines.py#L1
- https://github.com/python/cpython/blob/f8a79109d0c4f408d34d51861cc0a7c447f46d70/Lib/asyncio/windows_utils.py#L17

MichaReiser replied:
What happened to 

> There should be one-- and preferably only one --obvious way to do it.

AlexWaygood replied:
> What happened to
> 
> > There should be one-- and preferably only one --obvious way to do it.

many people are saying this

AlexWaygood replied:
Well, I'd quite like to use that `is_tuple_parenthesized` function you've got there in the formatter... what's the best course of action there? Copy-and-paste it, or move it to `ruff_python_trivia`?

MichaReiser replied:
You could add it to `ruff_python_ast` and either implement it as a free-standing function or as a method on `ExprTuple`.

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,823 @@
+}
+
+impl LineState {
+    fn receive_string_token(&mut self, token_value: String, token_range: TextRange) {
+        if self.first_item_in_line.is_none() {
+            self.first_item_in_line = Some((token_value, token_range));
+        } else {
+            self.following_items_in_line
+                .push((token_value, token_range));
+        }
+        self.comment_range_start = Some(token_range.end());
+    }
+
+    fn receive_comma_token(&mut self, token_range: TextRange) {
+        self.comment_range_start = Some(token_range.end());
+    }
+
+    fn receive_comment_token(&mut self, token_range: TextRange) {
+        self.comment_in_line = {
+            if let Some(range_start) = self.comment_range_start {
+                Some(TextRange::new(range_start, token_range.end()))
+            } else {
+                Some(token_range)
+            }
+        }
+    }
```
I wondered about using `visit_*` as names for these methods -- thoughts on that?

MichaReiser replied:
That would work for me.

AlexWaygood commented:
Okay, just pushed some major updates:
- The check is now implemented solely using the AST
- We now barely look at the raw tokens at all for fixing single-line `__all__` definitions.
- The machinery for multiline `__all__` definitions is much as it was, but I've tried to add more clarity in the code structure and comments.

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,883 @@
+    fn from(value: &str) -> Self {
+        Self {
+            category: InferredMemberType::of(value),
+            value: String::from(value),
```
This obviously isn't ideal :(

I wanted to make this struct generic over a lifetime, i.e.

```rs
struct AllItemSortKey<'a> {
    category: InferredMemberType,
    value: &'a str,
}
```

But had trouble making it work.

AlexWaygood commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,823 @@
+    following_items_in_line: Vec<(String, TextRange)>,
+    comment_range_start: Option<TextSize>,
+    comment_in_line: Option<TextRange>,
+}
```
I pushed a bunch more comments in my latest commits. But the key point here is that we're building up state that represents information on a single source line in a multiline `__all__` definition, _not_ a single "item" in a multiline `__all__` definition. We parse `__all__` in two stages:

1. We analyse each line of source code on its own, accumulating state until we reach the end of the line. This struct is basically a "bag of state" that collects that state until we get to the end of the line. Once we reach the end of the line, we use that state to produce a variant of the `DunderAllLine` enum (via `into_dunder_all_line()`).
2. Having constructed a vector of `DunderAllLine` variants, we then convert those into `DunderAllItem` variants.

AlexWaygood commented:
Here's the latest diff when running this autofix on CPython, FWIW: [cpython_diff.txt](https://github.com/astral-sh/ruff/files/13942003/cpython_diff.txt)

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+        _ => return,
+    };
+
+    let mut possibly_fixable = true;
+    let mut string_items = vec![];
+    for elt in elts {
+        // Don't flag `__all__` definitions that contain non-strings
+        let Some(string_literal) = elt.as_string_literal_expr() else {
+            return;
+        };
+        // If any strings are implicitly concatenated, don't bother trying to autofix
+        if string_literal.value.is_implicit_concatenated() {
+            possibly_fixable = false;
+        }
+        string_items.push(string_literal.value.to_str());
+    }
+    if dunder_all_is_already_sorted(&string_items) {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
```
It's hard to tell if it's worth it (probably not because `__all__` items are so uncommon). But we could avoid allocating a new `Vec` when the items are already sorted.


```suggestion
    let Some(unsorted) = UnsortedDunderAllItems::from_elements(elts) else {
        return;
    };

    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);

    if !unsorted.implicit_concatenated {
        if let Some(fix) = get_fix(range, elts, &unsorted.items, &kind, checker) {
            diagnostic.set_fix(fix);
        }
    }

    checker.diagnostics.push(diagnostic);
}

struct UnsortedDunderAllItems<'a> {
    items: Vec<&'a str>,
    implicit_concatenated: bool,
}

impl<'a> UnsortedDunderAllItems<'a> {
    fn from_elements(elements: &'a [Expr]) -> Option<Self> {
        let Some((first, rest)) = elements.split_first() else {
            return None;
        };

        if rest.is_empty() {
            return None;
        }

        let mut previous = first.as_string_literal_expr()?.value.to_str();

        for element in rest {
            let current = element.as_string_literal_expr()?.value.to_str();

            if AllItemSortKey::from(current) < AllItemSortKey::from(previous) {
                let mut implicit_concatenated = false;

                let mut items = Vec::with_capacity(elements.len());

                for element in elements {
                    let string_literal = element.as_string_literal_expr()?;
                    implicit_concatenated |= string_literal.value.is_implicit_concatenated();
                    items.push(string_literal.value.to_str());
                }

                return Some(Self {
                    items,
                    implicit_concatenated,
                });
            }

            previous = current;
        }

        // All Sorted
        None
    }
}

```

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
```
Nit: `create_fix`: Get is most often used to simply read a value (e.g. a field)

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+        if value.len() > 1 && is_cased_uppercase(value) {
+            Self::Constant
+        // E.g. `Class`
+        } else if value.chars().next().is_some_and(char::is_uppercase) {
```
Nit
```suggestion
        } else if value.starts_with(char::is_uppercase) {
```

AlexWaygood replied:
Ah, this logic was purely copied-and-pasted from the `isort` rule 😄

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+            locator,
+        );
+
+        self.items
+            .sort_by_cached_key(|item| AllItemSortKey::from(item));
```
Nit: We could consider to store `AllItemSortyKey` on the `DundarAllItem` if we're concerned about the cost. Although I don't think it's necessary because:

* This code path is only exercised for fixes, it's okay if they're somewhat slower
* We can change `AllItemSortKey` to us a `&str` to avoid cloning. The complexity then becomes `O(n)` but only for all capitalized strings.

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+    true
+}
+
+struct AllItemSortKey {
+    category: InferredMemberType,
+    value: String,
+}
+
```
We can avoid cloning here by using `self.items.sort_by` when sorting (I assume that's where you ran into lifetime issues)

```patch
Index: crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs	(revision d78e931ef1c7cf71946a4ce63ea5669aa2b27dac)
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs	(date 1705398024594)
-struct AllItemSortKey {
+struct AllItemSortKey<'a> {
     category: InferredMemberType,
-    value: String,
+    value: &'a str,
 }
 
-impl Ord for AllItemSortKey {
+impl Ord for AllItemSortKey<'_> {
     fn cmp(&self, other: &Self) -> Ordering {
         self.category
             .cmp(&other.category)
@@ -233,31 +249,31 @@
     }
 }
 
-impl PartialOrd for AllItemSortKey {
+impl PartialOrd for AllItemSortKey<'_> {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl PartialEq for AllItemSortKey {
+impl PartialEq for AllItemSortKey<'_> {
     fn eq(&self, other: &Self) -> bool {
         self.cmp(other) == Ordering::Equal
     }
 }
 
-impl Eq for AllItemSortKey {}
+impl Eq for AllItemSortKey<'_> {}
 
-impl From<&str> for AllItemSortKey {
-    fn from(value: &str) -> Self {
+impl<'a> From<&'a str> for AllItemSortKey<'a> {
+    fn from(value: &'a str) -> Self {
         Self {
             category: InferredMemberType::of(value),
-            value: String::from(value),
+            value,
         }
     }
 }
 
-impl From<&DunderAllItem> for AllItemSortKey {
-    fn from(item: &DunderAllItem) -> Self {
+impl<'a> From<&'a DunderAllItem> for AllItemSortKey<'a> {
+    fn from(item: &'a DunderAllItem) -> Self {
         Self::from(item.value.as_str())
     }
 }
@@ -434,7 +450,7 @@
         );
 
         self.items
-            .sort_by_cached_key(|item| AllItemSortKey::from(item));
+            .sort_by(|a, b| AllItemSortKey::from(a).cmp(&AllItemSortKey::from(b)));
         let joined_items = join_multiline_dunder_all_items(
             &self.items,
             locator,

```

AlexWaygood replied:
> We can avoid cloning here by using `self.items.sort_by` when sorting (I assume that's where you ran into lifetime issues)

That is indeed where I ran into lifetime issues, yup! TYVM

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+        }
+    };
+    let postlude = locator.slice(TextRange::new(postlude_start, dunder_all_range_end));
+    if !postlude.starts_with(newline) {
+        return Cow::Borrowed(postlude);
+    }
+    if TextSize::of(leading_indentation(postlude.trim_start_matches(newline)))
```
The `newline` returned by `Stylist` is the newline used for fixing, but it may or may not be the newline used in the `postlude`. For example, `Stylist` detects `\n` for a file with mixed line breaks if the first line uses `\n` but `postlude` could use `\r\n`, in which case this path incorrectly assumes that the postlude doesn't start with a newline. 

I don't understnad the code well enough to understand if that's a problem. If it is, changing it to `!postlude.starts_with(['\n', '\r'])` might be enough.

AlexWaygood replied:
Thanks, good call!

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+    locator: &Locator,
+) -> String {
+    let mut element_pairs = elts.iter().zip(elements).collect_vec();
+    element_pairs.sort_by_cached_key(|(_, elem)| AllItemSortKey::from(**elem));
```
Nit: Using `sort_by` is probably sufficient now that we can avoid allocating in `AllItemSortKey`

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -0,0 +1,882 @@
+    newline: &str,
+    needs_trailing_comma: bool,
+) -> String {
+    let max_index = sorted_items.len() - 1;
```
Nit: Rename to `last_item_index`. It holds slightly more information, because it gives meaning to what item the item with the `max` index points to.

MichaReiser reviewed (approved):
Overall this looks good to me, but there are a few clarification that are needed for me to understand the code and:

* We should align the `PartialEq` and `Ord` implementation of `AllDunderItems`
* Let's add some tests around unparenthesized tuples that start with a parenthesized first item (see inline comment). I suspect that things might go wrong in that case


It would be interesting to explore if we can implement the check as an AST rule and only use lexing for generating the fix. I'll leave that up to you if you want to tackle this and if so, if you want to do it as part of this PR.

MichaReiser reviewed (approved):
Nice improvements. I've a few suggestions but this looks good to me.

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs:
```diff
@@ -228,22 +217,57 @@ impl DunderAllKind<'_> {
+#[derive(Debug, is_macro::Is)]
+enum DunderAllSortClassification<'a> {
+    Sorted,
+    Invalid,
```
Nit: Maybe `Unsupported`. It's a valid `__all__` expression, but not one the fix supports

AlexWaygood replied:
Hmm, maybe `Unsortable` or `NotAStringLiteralList` would be best. We return this variant for truly invalid things like

```py
__all__ = [1, 2, 3]
```

 But we also return this variant for things that might or might not be valid, like

```py
foo = "whatever"
__all__ = [foo]
```

The thing that these have in common is that they're beyond the scope of this rule (not just the fix), so, while they might or might not be valid `__all__` definitions, we don't emit a violation for them

AlexWaygood replied:
I'll go with `Unsortable` 👍

AlexWaygood commented:
Thanks all for the reviews!
